### PR TITLE
fix: do not auto redirect to tls scheme

### PR DESCRIPTION
**Pull request by WarmSnowy**
This patch fix the problem that the client will
auto redirect to tls scheme even setting non-tls broker url which may confuse the user.

### CODE REVIEW

**Comment by WarmSnowy** [comment]
@tisonkun Hi, is there a plan to merge this patch?

File: src/service_discovery.rs [hunk 1]
```diff
@@ -116,13 +116,32 @@ impl<Exe: Executor> ServiceDiscovery<Exe> {
...
 
-            // use the TLS connection if available
-            let connection_url = if let Some(u) = &broker_url_tls {
-                u.clone()
-            } else if let Some(u) = &broker_url {
-                u.clone()
-            } else {
-                return Err(ServiceDiscoveryError::NotFound);
+            // Use broker url with the same schema of manager.url
+            let (connection_url, broker_url) = match self.manager.url.scheme() {
```
Comment (tisonkun) [inline src/service_discovery.rs]:
```suggestion
            let (connection_url, broker_url) = match base_url.scheme() {
```

**Review by tisonkun — APPROVED** [review]
Thanks for your contribution! This patch is good to go.

**Comment by tisonkun** [comment]
Although I wonder if you can add a function test to prevent the manner from being changed later, it's a bit difficult with our current codebase and test framework. I'll appreciate it a lot if you can achieve it while I don't make it a merge blocker.

Pending to merging once CI gives green.

File: src/service_discovery.rs [hunk 2]
```diff
@@ -116,13 +116,32 @@ impl<Exe: Executor> ServiceDiscovery<Exe> {
             } = convert_lookup_response(&response)?;
             is_authoritative = authoritative;
 
-            // use the TLS connection if available
-            let connection_url = if let Some(u) = &broker_url_tls {
```
Comment (cirias) [inline src/service_discovery.rs:120]: Why is `broker_url_tls` not None when the server doesn't support tls? Is it possible to be a misconfiguration?

File: src/service_discovery.rs [hunk 3]
```diff
@@ -116,13 +116,32 @@ impl<Exe: Executor> ServiceDiscovery<Exe> {
...
+                }
+                "pulsar" => {
+                    if let Some(u) = &broker_url {
+                        (
+                            u.clone(),
+                            format!("{}:{}", u.host_str().unwrap(), u.port().unwrap_or(6650)),
+                        )
+                    } else {
+                        return Err(ServiceDiscoveryError::NotFound);
+                    }
```
Comment (cirias) [inline src/service_discovery.rs]: nit: Not really a problem, but feel it's better to deduplicate the code here. Consider letting `match` return `(broker_url_maybe_tls, broker_port)`, then do `let connection_url = if let Some(..) = &broker_url_maybe_tls {` after.

File: src/service_discovery.rs (hunk 2 above)
Comment (WarmSnowy) [reply src/service_discovery.rs:120]: The `broker_url_tls` on here is parsed from server which the old client logic used to connect to server even if the configuration is non-tls url. It makes some trouble such as certificate verification problem even user not want to use tls.

Comment (cirias) [reply src/service_discovery.rs:120]: I was thinking if it's possible there is a server configuration causes the server to send the tls url even it doesn't handle tls. But that is not related to your change. All good, thank you!

**Review by cirias — COMMENTED** [review]
lgtm

**Comment by tisonkun** [comment]
Thanks for your contribution and review!

**Pull request merged**